### PR TITLE
Wrap response into error and result

### PR DESCRIPTION
**Pull request by donhardman**

### CODE REVIEW

File: manticoresearch_text_embeddings.h [hunk 1]
```diff
@@ -9,68 +9,48 @@
 #include <ostream>
 #include <new>
 
-using TextModelWrapper = void*;
+struct TextModelResult {
+  void *model;
```
Comment (klirichek) [inline manticoresearch_text_embeddings.h:13]: just to keep overall naming consistent:
`m_pModel` is more consistent in manticore style: m_ means 'member', p means 'pointer'

File: manticoresearch_text_embeddings.h [hunk 2]
```diff
@@ -9,68 +9,48 @@
...
+  char *error;
```
Comment (klirichek) [inline manticoresearch_text_embeddings.h:14]: m.b. `const char *m_szError` ? 
We don't expect c++ side will modify pointee. And 'sz' common for asciiz strings

File: manticoresearch_text_embeddings.h [hunk 3]
```diff
@@ -9,68 +9,48 @@
...
-           len == other.len &&
-           cap == other.cap;
-  }
-  bool operator!=(const FloatVec& other) const {
-    return ptr != other.ptr ||
-           len != other.len ||
-           cap != other.cap;
-  }
+struct FloatVecResult {
+  FloatVec vector;
```
Comment (klirichek) [inline manticoresearch_text_embeddings.h:34]: vector -> m_tData, or m_tEmbedding

File: manticoresearch_text_embeddings.h [hunk 4]
```diff
@@ -9,68 +9,48 @@
...
+  char *error;
```
Comment (klirichek) [inline manticoresearch_text_embeddings.h:35]: m_szError (as above)

File: manticoresearch_text_embeddings.h [hunk 5]
```diff
@@ -9,68 +9,48 @@
...
+
+using MakeVectEmbeddingsFn = FloatVecResult(*)(const TextModelWrapper*, const char*, uintptr_t);
 
-using DeleteVecFn = void(*)(FloatVec);
+using FreeVecResultFn = void(*)(FloatVecResult);
 
 using GetLenFn = uintptr_t(*)(const TextModelWrapper*);
 
 struct EmbedLib {
   uintptr_t size;
```
Comment (klirichek) [inline manticoresearch_text_embeddings.h:47]: here expected version tag.
size might be good when you constantly add new functions,
but actually that is version, or unique tag of the interface, that should be changed each time something published.

File: src/error.rs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+// Implement std::error::Error for LibError
+impl std::error::Error for LibError {}
+
+// Implement Display for LibError
+impl std::fmt::Display for LibError {
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		match self {
+			LibError::HuggingFaceApiBuildFailed => write!(f, "Failed to build Hugging Face API"),
```
Comment (sanikolaev) [inline src/error.rs:31]: This one looks confusing. Let's make it a little bit clearer.

File: src/error.rs [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+			LibError::ModelConfigFetchFailed => write!(f, "Failed to load model configuration"),
+			LibError::ModelConfigReadFailed => write!(f, "Failed to read model configuration"),
+			LibError::ModelConfigParseFailed => write!(f, "Failed to parse model configuration"),
+			LibError::ModelTokenizerFetchFailed => write!(f, "Failed to load model tokenizer"),
+			LibError::ModelTokenizerLoadFailed => write!(f, "Failed to load model tokenizer"),
+			LibError::ModelTokenizerConfigurationFailed => write!(f, "Failed to configure model tokenizer"),
+			LibError::ModelTokenizerEncodeFailed => write!(f, "Failed to encode text for model"),
+			LibError::ModelWeightsFetchFailed => write!(f, "Failed to load model weights"),
```
Comment (sanikolaev) [inline src/error.rs:39]:
```suggestion
			LibError::ModelWeightsFetchFailed => write!(f, "Failed to download model weights"),
```

File: src/error.rs [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+			LibError::ModelWeightsLoadFailed => write!(f, "Failed to load model weights"),
```
Comment (sanikolaev) [inline src/error.rs:40]:
```suggestion
			LibError::ModelWeightsLoadFailed => write!(f, "Failed to load model weights to memory"),
```

File: src/error.rs [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+// Implement Display for LibError
+impl std::fmt::Display for LibError {
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		match self {
+			LibError::HuggingFaceApiBuildFailed => write!(f, "Failed to build Hugging Face API"),
+			LibError::ModelConfigFetchFailed => write!(f, "Failed to load model configuration"),
+			LibError::ModelConfigReadFailed => write!(f, "Failed to read model configuration"),
+			LibError::ModelConfigParseFailed => write!(f, "Failed to parse model configuration"),
+			LibError::ModelTokenizerFetchFailed => write!(f, "Failed to load model tokenizer"),
```
Comment (sanikolaev) [inline src/error.rs:35]:
```suggestion
			LibError::ModelTokenizerFetchFailed => write!(f, "Failed to download model tokenizer"),
```

File: src/error.rs [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+			LibError::ModelTokenizerLoadFailed => write!(f, "Failed to load model tokenizer"),
```
Comment (sanikolaev) [inline src/error.rs:36]:
```suggestion
			LibError::ModelTokenizerLoadFailed => write!(f, "Failed to load model tokenizer to memory"),
```

File: src/error.rs [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+// Implement std::error::Error for LibError
+impl std::error::Error for LibError {}
+
+// Implement Display for LibError
+impl std::fmt::Display for LibError {
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		match self {
+			LibError::HuggingFaceApiBuildFailed => write!(f, "Failed to build Hugging Face API"),
+			LibError::ModelConfigFetchFailed => write!(f, "Failed to load model configuration"),
```
Comment (sanikolaev) [inline src/error.rs:32]:
```suggestion
			LibError::ModelConfigFetchFailed => write!(f, "Failed to download model configuration"),
```

File: src/error.rs [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+			LibError::ModelConfigReadFailed => write!(f, "Failed to read model configuration"),
+			LibError::ModelConfigParseFailed => write!(f, "Failed to parse model configuration"),
+			LibError::ModelTokenizerFetchFailed => write!(f, "Failed to load model tokenizer"),
+			LibError::ModelTokenizerLoadFailed => write!(f, "Failed to load model tokenizer"),
+			LibError::ModelTokenizerConfigurationFailed => write!(f, "Failed to configure model tokenizer"),
+			LibError::ModelTokenizerEncodeFailed => write!(f, "Failed to encode text for model"),
+			LibError::ModelWeightsFetchFailed => write!(f, "Failed to load model weights"),
+			LibError::ModelWeightsLoadFailed => write!(f, "Failed to load model weights"),
+			LibError::ModelLoadFailed => write!(f, "Failed to load model"),
```
Comment (sanikolaev) [inline src/error.rs:41]:
```suggestion
			LibError::ModelLoadFailed => write!(f, "Failed to load model to memory"),
```

**Review by sanikolaev — CHANGES_REQUESTED** [review]

**Pull request merged**